Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp

```diff
@@ -47,88 +48,353 @@
+        {
+            sendToMediaStreamTrackPrivate([] (auto& privateTrack) {
+                privateTrack.hasStartedProducingData();
+            });
+        }
+
+        void sendToMediaStreamTrackPrivate(Function<void(MediaStreamTrackPrivate&)>&& task)
+        {
+            m_postTask([task = WTFMove(task), privateTrack = m_privateTrack] () mutable {
+                if (RefPtr protectedPrivateTrack = privateTrack.get())
+                    task(*protectedPrivateTrack);
+            });
+        }
+
+        bool preventSourceFromEnding() { return m_shouldPreventSourceFromEnding; }
+
+        WeakPtr<MediaStreamTrackPrivate> m_privateTrack;
+        Ref<RealtimeMediaSource> m_source;
+        std::function<void(Function<void()>&&)> m_postTask;
+        bool m_shouldPreventSourceFromEnding { true };
+        bool m_isStarted { true };
+    };
+
+private:
+    MediaStreamTrackPrivateSourceObserverWrapper(Ref<RealtimeMediaSource>&& source, std::function<void(Function<void()>&&)>&& postTask)
+        : m_source(WTFMove(source))
+        , m_postTask(WTFMove(postTask))
+    {
+        ASSERT(m_postTask || isMainThread());
+        if (!m_postTask)
+            m_postTask = [] (Function<void()>&& function) { function(); };
+    }
+
+    Ref<RealtimeMediaSource> m_source;
+    std::unique_ptr<Observer> m_observer;
+    std::function<void(Function<void()>&&)> m_postTask;
+    HashMap<uint64_t, ApplyConstraintsHandler> m_applyConstraintsCallbacks;
+    uint64_t m_applyConstraintsCallbacksIdentifier { 0 };
+};
+
+MediaStreamTrackPrivate::MediaStreamTrackPrivate(Ref<const Logger>&& trackLogger, Ref<RealtimeMediaSource>&& source, String&& id, std::function<void(Function<void()>&&)>&& postTask)
+    : m_sourceObserver(MediaStreamTrackPrivateSourceObserverWrapper::create(WTFMove(source), WTFMove(postTask)))
     , m_id(WTFMove(id))
+    , m_label(m_sourceObserver->source().name())
+    , m_type(m_sourceObserver->source().type())
+    , m_deviceType(m_sourceObserver->source().deviceType())
+    , m_isCaptureTrack(m_sourceObserver->source().isCaptureSource())
+    , m_captureDidFail(m_sourceObserver->source().captureDidFail())
     , m_logger(WTFMove(trackLogger))
 #if !RELEASE_LOG_DISABLED
     , m_logIdentifier(uniqueLogIdentifier())
+#endif
+    , m_isProducingData(m_sourceObserver->source().isProducingData())
+    , m_isMuted(m_sourceObserver->source().muted())
+    , m_isInterrupted(m_sourceObserver->source().interrupted())
+    , m_settings(m_sourceObserver->source().settings())
+    , m_capabilities(m_sourceObserver->source().capabilities())
+#if ASSERT_ENABLED
+    , m_creationThreadId(isMainThread() ? 0 : Thread::current().uid())
 #endif
 {
-    ASSERT(isMainThread());
     UNUSED_PARAM(trackLogger);
     ALWAYS_LOG(LOGIDENTIFIER);
+
 #if !RELEASE_LOG_DISABLED
-    m_source->setLogger(m_logger.copyRef(), m_logIdentifier);
+    if (isMainThread())
+        m_sourceObserver->source().setLogger(m_logger.copyRef(), m_logIdentifier);
```

Comment:
Losing all logging will make debugging much more painful. Have you thought about how we might enable RealtimeMediaSource logging when a track is in a worker?

Reply:
All track sources are main thread only for now.
VideoTrackGenerator will be a new source and we should be able to set the logger directly based on the worker context.